source/Options.cpp

```diff
@@ -147,336 +147,176 @@ Options::Options(
-  }
+    model_generators_configuration_ =
+        parse_json_configuration_files(generator_configuration_paths_);
 
-  if (!variables["graphql-metadata-paths"].empty()) {
-    graphql_metadata_paths_ = check_path_exists(
-        variables["graphql-metadata-paths"].as<std::string>());
-  } else {
-    graphql_metadata_paths_ = "";
-  }
+    if (!json["model-generator-search-paths"].isNull()) {
+        model_generator_search_paths_ = parse_search_paths(
+            json["model-generator-search-paths"].asString());
+    }
 
-  if (!variables["proguard-configuration-paths"].empty()) {
-    proguard_configuration_paths_ = parse_paths_list(
-        variables["proguard-configuration-paths"].as<std::string>(),
-        /* extension */ ".pro");
-  }
+    repository_root_directory_ = check_directory_exists(
+        json["repository-root-directory"].asString());
+    source_root_directory_ = check_directory_exists(
+        json["source-root-directory"].asString());
 
-  if (!variables["generated-models-directory"].empty()) {
-    generated_models_directory_ = check_path_exists(
-        variables["generated-models-directory"].as<std::string>());
-  }
+    if (!json["source-exclude-directories"].isNull()) {
+        source_exclude_directories_ = parse_paths_list(
+            json["source-exclude-directories"].asString(),
+            /* extension */ std::nullopt,
+            /* check_exist */ false);
+    }
 
-  generator_configuration_paths_ = parse_paths_list(
-      variables["model-generator-configuration-paths"].as<std::string>(),
-      /* extension */ ".json");
-  model_generators_configuration_ =
-      parse_json_configuration_files(generator_configuration_paths_);
+    if (!json["grepo-metadata-path"].isNull()) {
+        grepo_metadata_path_ =
+            check_path_exists(json["grepo-metadata-path"].asString());
+    }
 
-  if (!variables["model-generator-search-paths"].empty()) {
-    model_generator_search_paths_ = parse_search_paths(
-        variables["model-generator-search-paths"].as<std::string>());
-  }
+    apk_path_ = check_path_exists(json["apk-path"].asString());
+    output_directory_ = std::filesystem::path(
+        check_directory_exists(json["output-directory"].asString()));
 
-  repository_root_directory_ = check_directory_exists(
-      variables["repository-root-directory"].as<std::string>());
-  source_root_directory_ = check_directory_exists(
-      variables["source-root-directory"].as<std::string>());
+    if (!json["sharded-models-directory"].isNull()) {
+        sharded_models_directory_ = std::filesystem::path(check_directory_exists(
+            json["sharded-models-directory"].asString()));
+    }
 
-  if (!variables["source-exclude-directories"].empty()) {
-    source_exclude_directories_ = parse_paths_list(
-        variables["source-exclude-directories"].as<std::string>(),
-        /* extension */ std::nullopt,
-        /* check_exist */ false);
-  }
+    sequential_ = json["sequential"].asBool();
```

Comment:
When `--sequential` is not provided to the command line in the shim, we don't even define a `sequential` key in json (see your code in the shim). In that case, I think `json["sequential"]` will be null. I'm surprised `asBool()` doesn't error on that.
Can we define a function `JsonValidation::optional_boolean(json, key, default_value)` and use it here and for the other booleans?